### DISCUSSION TITLE
Support renderToStringAsync for Solid JS

### DISCUSSION THREAD

**Discussion by patdx** [withastro/roadmap#519] [Proposal]
### Body

# Summary

Support async rendering for Solid JS.

# Background & Motivation

The Solid JS renderer uses renderToString, not renderToStringAsync. It seems a little out of place, considering that:

- [Other renderers such as Vue already use an async render function](https://github.com/withastro/astro/blob/main/packages/integrations/vue/server.js)
- Astro's `renderToStaticMarkup` already appears to support async functions / returned promises
- First party integrations like `@astrojs/image` provide async functions like `getImage()` that are not usable in sync-only Solid components

# Example

In my own project, I patched Astro Solid here: https://github.com/patdx/pmil.me/blob/33f26446c60456f8889402e2668f805f6ad229f6/.yarn/patches/%40astrojs-solid-js-npm-2.0.2-bccf2155a9.patch

And it worked quite nicely. Most notably, I was then able to build my own "async component" by using Suspense: https://github.com/patdx/pmil.me/blob/main/src/components/Image.tsx

The trickiest part of my experiment is that suspense boundaries and `createResource` calls seem to inject hydration scripts into the rendered HTML output, even for components that are not intended to hydrate. By wrapping the component in NoHydration and Suspense, it did not generate any unnecessary hydration scripts and it basically became an async component. I'm not sure if these components should be included by default or left up to users. Need to consider the impact on users of hydrated components too.

**Comment by natemoo-re** [comment]
This seems like a very reasonable change! Want to [make a PR](https://github.com/withastro/astro/pulls) directly?

**Reply by patdx** [reply]
Yeah sure, I'm giving it a try over here: https://github.com/patdx/astro/pull/1/files
Still need to investigate some different edge cases like timing out or throwing an async error.
I will submit it as a PR when it is more polished and I wrote some tests :sweat_smile:

**Reply by patdx** [reply]
I submitted a PR here: https://github.com/withastro/astro/pull/6791
